### PR TITLE
Add Initial SQL Server support

### PR DESCRIPTION
**Pull request by AdRiley**
### Pull Request Description

Adds initial SQL Server read support

image

There is still a lot to do on the dialect, but would prefer to move forwards with MRs that are less than 2000 lines of code. (Apologies that it is so big, I don't think it could have been much smaller.)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

**Review by Akirathan — APPROVED** [review]
Java/Scala changes are fine.

File: build.sbt [hunk 1]
```diff
@@ -163,7 +163,11 @@ GatherLicenses.distributions := Seq(
   makeStdLibDistribution("Database", Distribution.sbtProjects(`std-database`)),
   makeStdLibDistribution("Image", Distribution.sbtProjects(`std-image`)),
   makeStdLibDistribution("AWS", Distribution.sbtProjects(`std-aws`)),
-  makeStdLibDistribution("Snowflake", Distribution.sbtProjects(`std-snowflake`))
+  makeStdLibDistribution(
+    "Snowflake",
+    Distribution.sbtProjects(`std-snowflake`)
+  ),
+  makeStdLibDistribution("Microsoft", Distribution.sbtProjects(`std-microsoft`))
```
Comment (radeusgd) [inline build.sbt:170]: I assume this was discussed, so not forcing any changes, but it feels to me that a package called `Microsoft` may be too broad - they have **so many** different products that could fit into this library - I can't imagine as we progress and add more integrations with other MS products that they'd fit into a single library.

Was it considered to call this `MS_SQL_Server` or something like that? It feels like having smaller, more fine-grained packages is a safer approach.

File: distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso [hunk 2]
```diff
@@ -66,6 +66,19 @@ type Redshift_Dialect
...
+    get_limit_part_sql : Integer -> SQL_Builder
+    get_limit_part_sql self limit =
+        SQL_Builder.code (" LIMIT " + limit.to_text)
+
+    ## PRIVATE
+       Generates SQL for limiting the number of rows using the `TOP` clause.
+    get_top_part_sql : Integer -> SQL_Builder
+    get_top_part_sql self limit =
+        _ = [limit]
+        SQL_Builder.code ""
```
Comment (radeusgd) [inline distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso]: I don't like the fact that we have 2 functions here and one will always have to return empty code. IMO this is a bit of leaky abstraction - we assume there are exactly 2 ways to limit the query (probably true, I don't know a third way apart from TOP and LIMIT) - if we'd have to add another one we would need to modify all dialects again.

Could we merge this into a single method returning a pair - the SQL code and an `Integer` position where to put it - much like we do with the `context_extensions` mechanism that I introduced very recently? This would make it much more extensible.

Alternatively - if we actually want to assume there's only TOP and LIMIT - we could do a 'closed' approach - just do a simple getter on the dialect `query_limit_mode : Query_Limit_Mode` where `Query_Limit_Mode` would consist of 2 constructors: `Top` and `Limit`.

File: distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso [hunk 3]
```diff
@@ -139,6 +152,16 @@ type Redshift_Dialect
...
 
+    ## PRIVATE
+       Specifies whether the Database supports CREATE TEMPORARY TABLE syntax.
+    suppports_temporary_table_syntax : Boolean
+    suppports_temporary_table_syntax self = True
+
+    ## PRIVATE
+       Specifies whether the Database supports CREATE TABLE #temptablename syntax.
+    suppports_temporary_with_hash_syntax : Boolean
+    suppports_temporary_with_hash_syntax self = False
```
Comment (radeusgd) [inline distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso]: Again, it feels to me like this is a bit of duplication. Is there a database that actually supports both syntaxes?

I'd prefer to have `temporary_table_syntax -> Temporary_Table_Syntax` where that type could have e.g. ~3 variants: `Temporary_Keyword`, `Hash_Syntax`, `Not_Supported`.

Comment (radeusgd) [reply distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso]: Also, I don't see `suppports_temporary_with_hash_syntax` used anywhere? I assume this is deliberate and will come in some future PR?

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso [hunk 4]
```diff
@@ -590,7 +594,7 @@ generate_create_table dialect name columns primary_key temporary =
     column_definitions = columns.map (generate_column_description dialect)
     modifiers = if primary_key.is_nothing then [] else
         [SQL_Builder.code ", PRIMARY KEY (" ++ SQL_Builder.join ", " (primary_key.map dialect.wrap_identifier) ++ ")"]
-    table_type = if temporary then "TEMPORARY TABLE" else "TABLE"
+    table_type = if temporary && dialect.suppports_temporary_table_syntax then "TEMPORARY TABLE" else "TABLE"
```
Comment (radeusgd) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso:590]: It feels to me that we should fail saying that this is not supported. It may trip too many tests here, but maybe at least a warning??

I'm feeling rather anxious that we are telling the user that we are creating a temporary table but it is actually a permanent one - this can very easily lead to users polluting their DBs with tons of garbage tables by accident.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Connection.enso [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+       Creates a SQLServer connection based on a URL and properties.
+
+       Arguments:
+       - url: The URL to connect to.
+       - properties: A vector of properties for the connection.
+       - make_new: A function that returns a new connection.
+    create : Text -> Vector -> (Text -> Text -> SQLServer_Connection) -> SQLServer_Connection
+    create url properties make_new =
+        jdbc_connection = JDBC_Connection.create url properties
+        entity_naming_properties = Entity_Naming_Properties.from_jdbc_connection jdbc_connection is_case_sensitive=False
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Connection.enso:36]: So T-SQL is truly case insensitive, interesting.

It would be nice to create a test in which we do `(Table.new [["A", [1]], ["a", [2]]]).select_into_database_table ...` and see what happens.

Comment (radeusgd) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Connection.enso:36]: Of course we can add it in next iterations, just noting so that we can keep track of that.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Connection.enso [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+
+    ## ICON metadata
+       Returns the name of the current warehouse.
+    warehouse : Text
+    warehouse self =
+        vector = self.connection.read_single_column "SELECT current_warehouse() as name" "name"
+        vector.first
+
+    ## ICON data_input
+       Returns a new Connection with the specified database set as default.
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Connection.enso]:
```suggestion
       Returns a new Connection with the specified warehouse set as default.
```

Should this comment be updated?

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/SQLServer_Details.enso [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    ## PRIVATE
+       Build the Connection resource.
+
+       Arguments:
+       - options: Overrides for the connection properties.
+    connect : Connection_Options -> SQLServer_Connection
+    connect self options =
+        properties = options.merge self.jdbc_properties
+
+        ## Cannot use default argument values as gets in an infinite loop if you do.
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/SQLServer_Details.enso]: Is this comment still relevant? I'm not sure if I understand why an infinite loop would happen here.

File: std-bits/database/src/main/java/org/enso/database/JDBCUtils.java [hunk 8]
```diff
@@ -70,9 +72,22 @@ public static void setLocalTime(PreparedStatement stmt, int columnIndex, LocalTi
...
+   * Sets a LocalTime in a PreparedStatement via TimeStamp.
+   *
+   * @param stmt
+   */
+  public static void setLocalTimeViaTimeStamp(
+      PreparedStatement stmt, int columnIndex, LocalTime localTime) throws SQLException {
+
+    Timestamp timestamp = Timestamp.valueOf(localTime.atDate(LocalDate.of(1970, 1, 1)));
+    stmt.setTimestamp(columnIndex, timestamp);
+  }
```
Comment (radeusgd) [inline std-bits/database/src/main/java/org/enso/database/JDBCUtils.java:85]: Won't this lose µs and ns parts?

File: test/Microsoft_Tests/src/SQLServer_Spec.enso [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        t = connection.query (SQL_Query.Table_Name tinfo)
+        row1 = ["a", Nothing, False, 1.2, 1.2, 0, 0, 0, Time_Of_Day.new 12 12 12 1 1 1, Date.new 2021 1 1, Date_Time.new 2021 1 1 12 12 12 500 1 1, Date_Time.new 2021 1 1 12 12 12 1 1 1, Date_Time.new 2021 1 1 12 12 12 1 1 1, Date_Time.new 2021 1 1 12 12 12 1 1 1]
+        row2 = ["abc", Nothing, Nothing, 1.3, 1.3, 255, 32767, 9223372036854775807, Time_Of_Day.new 7 12 12 1 1 1, Date.new 1999 1 1, Date_Time.new 1999 1 1 12 12 12 1 1 1, Date_Time.new 1999 1 1 12 12 12 1 1 1, Date_Time.new 1999 1 1 12 12 12 1 1 1, Date_Time.new 1999 1 1 12 12 12 1 1 1]
+        row3 = ["def", 42, True, 1.4, 1.4, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing]
+        table = Table.from_rows ["strs", "ints", "bools", "reals", "floats", "tinyints", "smallints", "bigints", "times", "dates", "datetimes", "smalldatetimes", "datetime2s", "datetimeoffsets"] [row1, row2, row3]
+        table2 = table.cast ['tinyints'] (Value_Type.Integer Bits.Bits_16)
+        table3 = table2.cast ['smallints'] (Value_Type.Integer Bits.Bits_16)
+        table4 = table3.cast ['ints'] (Value_Type.Integer Bits.Bits_32)
```
Comment (radeusgd) [inline test/Microsoft_Tests/src/SQLServer_Spec.enso]: I'd suggest
```suggestion
        source_table = Table.from_rows ["strs", "ints", "bools", "reals", "floats", "tinyints", "smallints", "bigints", "times", "dates", "datetimes", "smalldatetimes", "datetime2s", "datetimeoffsets"] [row1, row2, row3]
            . cast ['tinyints'] (Value_Type.Integer Bits.Bits_16)
            . cast ['smallints'] (Value_Type.Integer Bits.Bits_16)
            . cast ['ints'] (Value_Type.Integer Bits.Bits_32)
```

File: test/Microsoft_Tests/src/SQLServer_Spec.enso [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        _ -> Database.connect connection_details
+
+add_specs suite_builder =
+    case create_connection_builder of
+        Nothing ->
+            message = "SQLServer test database is not configured. See README.md for instructions."
+            suite_builder.group "[SQLServer] Database tests" pending=message (_-> Nothing)
+        connection_builder ->
+            suite_builder.group "[SQLServer] Small" group_builder->
+                group_builder.specify "debugtest" <|
```
Comment (radeusgd) [inline test/Microsoft_Tests/src/SQLServer_Spec.enso]: can we have a more meaningful name here?

File: test/Microsoft_Tests/src/SQLServer_Spec.enso [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+                    connection = Database.connect get_configured_connection_details
+                    tinfo = Name_Generator.random_name "Tinfo"
+                    connection.execute_update 'Create Table "'+tinfo+'" ("times" TIME)'
+                    t = connection.query (SQL_Query.Table_Name tinfo)
+                    row1 = [Time_Of_Day.new 12 12 12 1 1 1]
+                    row2 = [Time_Of_Day.new 7 12 12 1 1 1]
+                    row3 = [ Nothing]
+                    table = Table.from_rows ["times"] [row1, row2, row3]
+                    Panic.rethrow <|
+                        t.update_rows table update_action=Update_Action.Insert
```
Comment (radeusgd) [inline test/Microsoft_Tests/src/SQLServer_Spec.enso]: We should try to clean-up this table.

Our server runs in Docker image so it is temporary. But if someone runs it on some permanent local deployment, they will be getting more and more trash tables with each run.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+polyglot java import java.sql.Types
+
+## PRIVATE
+type SQLServer_Type_Mapping
+    ## PRIVATE
+    value_type_to_sql : Value_Type -> Problem_Behavior -> SQL_Type
+    value_type_to_sql value_type on_problems =
+        result = case value_type of
+            Value_Type.Boolean -> SQL_Type.Value Types.BIT "BIT"
+            ## A SQLServer TinyInt is untyped with a range of 0-255,
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso]: Did you mean unsigned?
```suggestion
            ## A SQLServer TinyInt is unsigned with a range of 0-255, 
```

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+                    "An unsupported SQL type ["+type_name.to_text+"] cannot be converted into an SQL type because it did not contain the SQL metadata needed to reconstruct it."
+
+        approximated_value_type = SQLServer_Type_Mapping.sql_type_to_value_type result
+        problems = if approximated_value_type == value_type then [] else [Inexact_Type_Coercion.Warning value_type approximated_value_type]
+        on_problems.attach_problems_before problems result
+
+    ## PRIVATE
+    sql_type_to_value_type : SQL_Type -> Value_Type
+    sql_type_to_value_type sql_type =
+        simple_type = case sql_type.typeid of
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso]: This looks so much cleaner than the Dictionaries we had. Thanks @JaroslavTulach for #7607 🎉 

We probably should restructure Postgres/Snowflake to also prefer this approach.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            Types.BIT -> Value_Type.Boolean
+            Types.TINYINT -> Value_Type.Integer Bits.Bits_16
+            Types.SMALLINT -> Value_Type.Integer Bits.Bits_16
+            Types.INTEGER -> Value_Type.Integer Bits.Bits_32
+            Types.BIGINT -> Value_Type.Integer Bits.Bits_64
+            ## SQLServer real is 32-bit, but our In Memory engine only has 64-bit floating point numbers so we have to upcast it.
+            Types.REAL -> Value_Type.Float Bits.Bits_64
+            Types.DOUBLE -> Value_Type.Float Bits.Bits_64
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso]: This is not the right approach.

A dialect's type mapping deals with how the SQL types map to our `Value_Type` representation of the same concept. They should reflect what is really there in the Database. So if the Database supports 32-bit floats, they should map to `Value_Type.Float ..Bits_32` - that is exactly the purpose of that type.

In the current state - if I load in a SQL Server table that has a `REAL` column - it will wrongly report its type as 64-bit float - but it really is 32-bit.

This is the mapping of types itself. Now, how these types are handled during materialization to memory is a different story - I think our `double_fetcher` will automatically materialize either variant of `Float` as 64-bit floats that are the only ones we currently support in-memory.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+            _ -> case sql_type.name of
+                "datetimeoffset" -> Value_Type.Date_Time with_timezone=True
+                _ -> Nothing
+        simple_type.if_nothing <|
+            ## If we didn't match any of the types from the simple mapping, we
+               continue with the more complex mappings that take stuff like
+               precision into account.
+            case complex_types_map.get sql_type.typeid Nothing of
+                Nothing -> on_unknown_type sql_type
+                builder -> builder sql_type
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso]: I think that with the `case-of` pattern match here, we no longer need to distinguish between the _"simple"_ and _"complex"_ mappings - we should probably merge all this into a big 'switch' statement. Before this was split only because we could not use the Java enum constants in the pattern match and needed various hacks to make the code readable.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+            # We currently do not support bit types.
+            on_unknown_type sql_type
+    handle_timestamp sql_type = case sql_type.name of
+        "TIMESTAMPTZ" -> Value_Type.Date_Time with_timezone=True
+        "TIMESTAMP_TZ" -> Value_Type.Date_Time with_timezone=True
+        "TIMESTAMPLTZ" -> Value_Type.Date_Time with_timezone=False
+        "TIMESTAMP_LTZ" -> Value_Type.Date_Time with_timezone=False
+        "TIMESTAMPNTZ"   -> Value_Type.Date_Time with_timezone=False
+        "TIMESTAMP_NTZ"   -> Value_Type.Date_Time with_timezone=False
+        _             -> on_unknown_type sql_type
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso]: After a brief look at T-SQL data types, this seems off. Isn't this a copy-paste from Snowflake?

I couldn't find these names used in T-SQL - it rather uses `DATETIMEOFFSET`/`DATETIME2` like you used above.

OK to address later.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso [hunk 17]
```diff
@@ -0,0 +1,720 @@
...
+    Value dialect_operations
+
+    ## PRIVATE
+       Name of the dialect.
+    name : Text
+    name self = sqlserver_dialect_name
+
+    ## PRIVATE
+    to_text : Text
+    to_text self = "SQLServer_Dialect"
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso]: I'm wondering if it would make sense to call this `Transact_SQL_Dialect`? Or is there a subtle difference between T-SQL and the specifics of SQL Server?

File: build.sbt (hunk 1 above)
Comment (AdRiley) [reply build.sbt:170]: I think I started with SQLServer and @jdunkerley convinced me it should be Microsoft

Comment (jdunkerley) [reply build.sbt:170]: We have AWS and Google. Having the full set of the big three feels correct, so Standard.Microsoft. We considered moving Excel but it is so core its almost like plain text must be support by default.

Yes I expect all of them will end up with sub-libraries that you include what you need one day.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Connection.enso (hunk 6 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Connection.enso]: No such warehouse in SQL Server - guessing this came from snowflake

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso (hunk 17 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso]: Not that I expect we will do it anytime soon but Sybase has a different Transact SQL dialect. I think SQLServer_Dialect is fine but maybe should be SQL_Server.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/SQLServer_Details.enso (hunk 7 above)
Comment (AdRiley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/SQLServer_Details.enso]: Not sure. It is copied pasted... Postgres and Snowflake both have the same comment. Not that that makes it correct :)

Comment (AdRiley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/SQLServer_Details.enso]: Removed it from here

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso (hunk 16 above)
Comment (AdRiley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso]: Yes left over from the snowflake starting point :)

Comment (AdRiley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso]: Removed

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso (hunk 14 above)
Comment (AdRiley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso]: Fixed

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso (hunk 15 above)
Comment (AdRiley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso]: Done

File: std-bits/database/src/main/java/org/enso/database/JDBCUtils.java (hunk 8 above)
Comment (AdRiley) [reply std-bits/database/src/main/java/org/enso/database/JDBCUtils.java:85]: I don't think so. I added a test "test datetime2 precision round trip"

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Connection.enso (hunk 5 above)
Comment (AdRiley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Connection.enso:36]: https://github.com/enso-org/enso/issues/10710

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso (hunk 4 above)
Comment (AdRiley) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso:590]: Will fix as a quick follow up https://github.com/enso-org/enso/issues/10711

File: distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso (hunk 3 above)
Comment (AdRiley) [reply distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso]: Removed suppports_temporary_with_hash_syntax for now. Will refactor as part of https://github.com/enso-org/enso/issues/10711

File: distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso (hunk 2 above)
Comment (AdRiley) [reply distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso]: Done

File: build/build/src/enso.rs [hunk 18]
```diff
@@ -145,7 +145,12 @@ impl BuiltEnso {
         let may_need_postgres = match &test_selection {
             StandardLibraryTestsSelection::All => true,
             StandardLibraryTestsSelection::Selected(only) =>
-                only.iter().any(|test| test.contains("Postgres_Tests")),
+                only.iter().any(|test| test.contains("Table_Tests")),
+        };
+        let may_need_sqlserver = match &test_selection {
+            StandardLibraryTestsSelection::All => true,
+            StandardLibraryTestsSelection::Selected(only) =>
+                only.iter().any(|test| test.contains("SQLServer")),
```
Comment (radeusgd) [inline build/build/src/enso.rs]: I think this is not right, the project is called `Microsoft_Tests`, right?

```suggestion
                only.iter().any(|test| test.contains("Microsoft_Tests")),
```

File: build/build/src/enso.rs [hunk 19]
```diff
@@ -198,7 +203,6 @@ impl BuiltEnso {
             _ => None,
         };
 
-        let std_tests = crate::paths::discover_standard_library_tests(&paths.repo_root)?;
```
Comment (radeusgd) [inline build/build/src/enso.rs]: If this line disappeared, how does the next line still compile??

File: distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso [hunk 20]
```diff
@@ -67,17 +67,10 @@ type Redshift_Dialect
...
-        SQL_Builder.code (" LIMIT " + limit.to_text)
-
-    ## PRIVATE
-       Generates SQL for limiting the number of rows using the `TOP` clause.
-    get_top_part_sql : Integer -> SQL_Builder
-    get_top_part_sql self limit =
-        _ = [limit]
-        SQL_Builder.code ""
+       Generates SQL modifier for limiting the number of rows and its position in the query
+    get_limit_sql_modifier : Integer -> Any
```
Comment (radeusgd) [inline distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso:71]: I'd prefer if we could make the type more precise. Could this be just `Context_Extension`?

Comment (radeusgd) [reply distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso:71]: But I guess the pair is OK too, it would just be good to document somewhere what should be put there so that future implementors do not have to rely solely on examples and usage,

File: distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso [hunk 21]
```diff
@@ -304,8 +304,8 @@ type Value_Type
             Value_Type.Float _ -> True
             Value_Type.Decimal _ _ -> True
             _ -> False
-        Value_Type.Float self_bits -> case target_type of
-            Value_Type.Float target_bits -> target_bits >= self_bits
+        Value_Type.Float _ -> case target_type of
+            Value_Type.Float _ -> True
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso]: Shall we add a comment that this is valid until we get `f32` support on in-memory? I assume once that happens, we probably want to revert this change.

We can even link a ticket: https://github.com/enso-org/enso/issues/6109

File: test/Microsoft_Tests/src/SQLServer_Spec.enso [hunk 22]
```diff
@@ -137,12 +126,24 @@ add_specs suite_builder =
...
+                    source_table = Table.from_rows ["dt2"] [row1, row2, row3]
+                    t.update_rows source_table update_action=Update_Action.Insert
+                    ## SQLServer only supports precision to 100 nanoseconds
+                    expected_row1 = [Date_Time.new 2021 1 1 12 13 14 500 1 0]
+                    expected_row2 = [Date_Time.new 2021 1 1 9 12 12 987 654 300]
+                    expected_row3 = [Nothing]
+                    expected_table = Table.from_rows ["dt2"] [expected_row1, expected_row2, expected_row3]
+                    returned_table = t.read
+                    returned_table.should_equal expected_table
+                    data.connection.execute_update 'DROP TABLE "'+name+'"'
```
Comment (radeusgd) [inline test/Microsoft_Tests/src/SQLServer_Spec.enso:146]: This will not execute if the test fails, leaving trash in the table.

While the CI runs it as docker image, some users may use a persistent DB - we should generally try to leave a clean env afterwards.

Can we just make this a temporary table?? If not, I'd do the DROP inside of `Panic.with_finalizer`

**Review by radeusgd — APPROVED** [review]
Looks good, just a few small nits

File: distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso (hunk 20 above)
Comment (AdRiley) [reply distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso:71]: I'd like to revisit the approach at some point. I am unsure about the "magic numbers" for the position in the SQL.

File: build/build/src/enso.rs (hunk 19 above)
Comment (AdRiley) [reply build/build/src/enso.rs]: Having weird merge issues with this line. I will put it back.

Comment (AdRiley) [reply build/build/src/enso.rs]: Oh you moved it to line 139 in a previous MR

File: distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso (hunk 21 above)
Comment (AdRiley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso]: I don't know if we want to revert it. An automatic cast with a warning seems more useful than an error telling me I have to do a manual cast. I'll add a comment to think about it.

File: test/Microsoft_Tests/src/SQLServer_Spec.enso (hunk 22 above)
Comment (AdRiley) [reply test/Microsoft_Tests/src/SQLServer_Spec.enso:146]: I'll make it a temporary after SQL Server has support for them which will be my next ticket. Had trouble doing it manually.

Comment (radeusgd) [reply test/Microsoft_Tests/src/SQLServer_Spec.enso:146]: Ok, sure

File: distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso (hunk 21 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso]: Hmm, I think you are right.

I'm not a fan of this because this is a **write** operation, that cannot be easily undone. So if the user sees the warning it is already too late to address it, all they can do is acknowledge it.

In this case though, in most cases the user would indeed cast as there isn't much they can do. And float is an inexact type anyway - so I guess the (additional) loss of precision is not that big of a deal here as it would be elsewhere.

So happy to keep it indeed.

Comment (AdRiley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso]: But with dry run mode the user gets to see the warning before they make the write.

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso]: Fair enough! I forgot about that case.

File: build/build/src/sqlserver.rs [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+        env::tests::ENSO_SQLSERVER_PASSWORD.remove();
+    }
+
+    pub async fn cleanup(&self) -> Result {
+        Docker.remove_container(&self.sqlserver_container, true).await
+    }
+}
+
+/// Retrieve input from asynchronous reader line by line and feed them into the given function.
+pub async fn process_lines<R: AsyncRead + Unpin>(reader: R, f: impl Fn(String)) -> Result<R> {
```
Comment (Frizi) [inline build/build/src/sqlserver.rs:113]: This function looks to be unused.

**Review by Frizi — APPROVED** [review]

**Review by hubertp — APPROVED** [review]
